File: ios/VineTrack/App/ChemicalIntelligence/ChemicalDefaultRateValidity.swift
import Foundation

/// Whether a stored operational default may be believed, and what amount it
/// actually records.
///
/// The iOS counterpart of Android's `ChemicalDefaultRateValidity`. The two are
/// kept deliberately identical: a rate that is spray-ready on one platform and
/// silently ignored on the other is a worse failure than either behaviour on
/// its own, because the disagreement is invisible until a tank is mixed.
///
/// Nothing here reads `rate_per_ha`. That column is a legacy projection with no
/// link back to a printed direction (sql/222), so it can never establish what
/// the operator confirmed.
nonisolated enum ChemicalDefaultRateValidity {

    /// Rate units a line may carry. Keyed lowercase; the value is the canonical
    /// spelling. These are LABEL rate units, never pack units.
    private static let supportedUnits: [String: String] = [
        "l": "L", "litre": "L", "litres": "L", "liter": "L", "liters": "L",
        "ml": "mL", "millilitre": "mL", "millilitres": "mL",
        "kg": "kg", "kilogram": "kg", "kilograms": "kg",
        "g": "g", "gram": "g", "grams": "g"
    ]

    /// Canonical spelling of a rate unit, or `nil` when it is not one a line
    /// may carry.
    static func canonicalUnit(_ raw: String?) -> String? {
        let key = raw?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() ?? ""
        return supportedUnits[key]
    }

    /// The two amount shapes a slot may hold — never both (shared shape D3).
    nonisolated enum Amount: Equatable, Sendable {
        /// A confirmed dose. The only shape that may prefill a spray line.
        case scalar(Double)
        /// What the label permits. NOT a decision, and never a prefill.
        case range(min: Double, max: Double)
    }

    private static func isUsableNumber(_ value: Double?) -> Bool {
        guard let value else { return false }
        return value.isFinite && value > 0
    }

    private static func amount(of slot: StoredChemicalDefaultRate) -> Amount? {
        let hasScalar = slot.value != nil
        let hasMin = slot.minValue != nil
        let hasMax = slot.maxValue != nil

        if hasScalar, !hasMin, !hasMax {
            guard let value = slot.value, isUsableNumber(value) else { return nil }
            return .scalar(value)
        }
        if !hasScalar, hasMin, hasMax {
            guard let min = slot.minValue, let max = slot.maxValue,
                  isUsableNumber(min), isUsableNumber(max),
                  // An inverted band is not a narrow band, it is a corrupt one.
                  min <= max
            else { return nil }
            return .range(min: min, max: max)
        }
        // Everything else: a lone bound, an empty row, or both shapes at once.
        return nil
    }

    /// A stored slot that passed every rule, with its amount already resolved.
    nonisolated struct ValidSlot: Equatable, Sendable {
        let basis: ChemicalDefaultRateBasis
        /// Canonical spelling of the LABEL rate's unit — never the pack unit.
        let unit: String
        let amount: Amount
        let slot: StoredChemicalDefaultRate

        /// The confirmed dose, or `nil` when this slot records an unnarrowed band.
        var scalar: Double? {
            if case .scalar(let value) = amount { return value }
            return nil
        }

        /// The registered band, or `nil` when this slot records a single dose.
        var range: (min: Double, max: Double)? {
            if case .range(let min, let max) = amount { return (min, max) }
            return nil
        }
    }

    /// The slot stored on `basis`, or `nil` when there is none the contract can
    /// believe.
    ///
    /// Checked in full, because each rule catches a different way a row stops
    /// meaning what it says:
    ///
    /// ```text
    /// root version          a document written to a contract this build cannot read
    /// containing basis      a per-100 L rate filed under per-hectare, applied per hectare
    /// option key identity   a key no minter produced, so no client can match it
    /// rate ids present      a default citing nothing can never be shown to be current
    /// rate id identity      a UUID cites a row, not a printed direction
    /// supported unit        an amount nothing can display or cost
    /// source vocabulary     provenance outside the closed set is unattributable
    /// amount shape (D3)     scalar XOR range, both finite and positive
    /// ```
    static func validSlot(
        _ defaults: StoredChemicalDefaultRates?,
        basis: ChemicalDefaultRateBasis
    ) -> ValidSlot? {
        guard let defaults else { return nil }
        guard defaults.version == StoredChemicalDefaultRates.currentVersion else { return nil }
        guard let slot = defaults.slot(basis) else { return nil }

        let key = slot.optionKey.trimmingCharacters(in: .whitespacesAndNewlines)
        let keyPrefix = ChemicalServerDefaultRateOption.optionKeyPrefix
        guard key.hasPrefix(keyPrefix), key.count > keyPrefix.count else { return nil }

        // Trimmed rather than filtered: a blank entry in the citation list is a
        // malformed row, not a row with one fewer citation.
        let idPrefix = ChemicalServerDefaultRateOption.rateIDPrefix
        let ids = slot.rateIds.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
        guard !ids.isEmpty else { return nil }
        guard ids.allSatisfy({ $0.hasPrefix(idPrefix) && $0.count > idPrefix.count }) else {
            return nil
        }

        guard slot.basis.trimmingCharacters(in: .whitespacesAndNewlines) == basis.rawValue else {
            return nil
        }
        guard let unit = canonicalUnit(slot.unit) else { return nil }

        let source = slot.source.trimmingCharacters(in: .whitespacesAndNewlines)
        let knownSource = source == StoredChemicalDefaultRate.sourceOperator
            || source == StoredChemicalDefaultRate.sourceRecommended
        guard knownSource else { return nil }

        guard let amount = amount(of: slot) else { return nil }
        return ValidSlot(basis: basis, unit: unit, amount: amount, slot: slot)
    }

    /// The CONFIRMED dose stored on `basis`, or `nil` when there is none.
    ///
    /// The scalar shape is the only confirmed amount: a slot still holding a
    /// band is a decision that was never finished, and its bounds are what the
    /// label permits rather than what this vineyard pours.
    static func confirmedScalar(
        _ defaults: StoredChemicalDefaultRates?,
        basis: ChemicalDefaultRateBasis
    ) -> ValidSlot? {
        guard let slot = validSlot(defaults, basis: basis), slot.scalar != nil else { return nil }
        return slot
    }

    /// Every basis carrying a believable slot, per-hectare first.
    static func validSlots(_ defaults: StoredChemicalDefaultRates?) -> [ValidSlot] {
        [
            validSlot(defaults, basis: .perHectare),
            validSlot(defaults, basis: .per100Litres)
        ].compactMap { $0 }
    }
}

extension StoredChemicalDefaultRates {
    /// The confirmed single dose on `basis`, or `nil` when this chemical records
    /// a range, records nothing, or records something malformed.
    ///
    /// This is the accessor the legacy `rate_per_ha` projection is built from,
    /// which is why it must never fall back to a bound or a midpoint.
    nonisolated func confirmedScalarAmount(for basis: ChemicalDefaultRateBasis) -> Double? {
        ChemicalDefaultRateValidity.confirmedScalar(self, basis: basis)?.scalar
    }

    /// The believable slot on `basis`, scalar or range.
    nonisolated func validSlot(for basis: ChemicalDefaultRateBasis) -> ChemicalDefaultRateValidity.ValidSlot? {
        ChemicalDefaultRateValidity.validSlot(self, basis: basis)
    }
}

File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/chemical/ChemicalDefaultRateValidity.kt
package com.rork.vinetrack.data.chemical

/**
 * Whether a PERSISTED `default_rates` slot may be believed.
 *
 * # Why a validator exists at all
 *
 * `default_rates` is written by three clients and read by four screens. Until
 * this object existed each reader made its own decision about what counted as
 * usable, and the loosest one won: Android accepted any positive `value`, so a
 * row that the Portal and the edge function both rejected as malformed still
 * prefilled a spray line here. A rate that one client refuses to display must
 * never be a rate another client puts in the tank, so the shape test lives in
 * ONE place and both the store display and the spray handoff ask it.
 *
 * Mirrors `isUsableStoredDefault` in
 * `supabase/functions/chemical-info-lookup/default_rates.ts`, rule for rule.
 *
 * # Nothing here repairs anything
 *
 * Every rule below answers yes or no. None of them mends a value, infers a
 * missing bound, re-mints an identity or substitutes a replacement rate. A
 * malformed default means the operator has not got a usable confirmed rate,
 * which is a true and recoverable state — they confirm one again. Quietly
 * "fixing" the row would manufacture a decision nobody made, and it would do
 * so at the exact point where the number is about to be poured.
 */
object ChemicalDefaultRateValidity {

    /** Every option key the identity minter produces carries this prefix. */
    const val OPTION_KEY_PREFIX: String = "${ChemicalDefaultRateIdentity.OPTION_ID_VERSION}_"

    /**
     * Gate D1 rate identities. A UUID here means the row cites a row-id rather
     * than a printed direction, which is precisely the untraceable provenance
     * the structured contract replaced.
     */
    const val RATE_ID_PREFIX: String = "rate_v1_"

    /**
     * The units a stored default may be expressed in.
     *
     * Closed on purpose, and the single source for both readers: a unit
     * nothing recognises cannot be displayed, costed or compared safely.
     */
    private val supportedUnits: Map<String, String> = mapOf(
        "l" to "L",
        "litre" to "L",
        "litres" to "L",
        "liter" to "L",
        "liters" to "L",
        "ml" to "mL",
        "millilitre" to "mL",
        "millilitres" to "mL",
        "kg" to "kg",
        "kilogram" to "kg",
        "kilograms" to "kg",
        "g" to "g",
        "gram" to "g",
        "grams" to "g",
    )

    /**
     * Canonical spelling of a rate unit, or null when it is not one a stored
     * default may carry.
     *
     * Normalises SPELLING AND CASE ONLY. It never converts: `g` stays `g` and
     * is never restated as `0.001 kg`, because the confirmed amount and its
     * unit are the operator's decision exactly as they made it.
     */
    fun canonicalUnit(raw: String?): String? =
        supportedUnits[raw?.trim()?.lowercase().orEmpty()]

    /** The two amount shapes a slot may hold — never both (shared shape D3). */
    sealed interface Amount {
        /** A confirmed dose. The only shape that may prefill a spray line. */
        data class Scalar(val value: Double) : Amount

        /** What the label permits. NOT a decision, and never a prefill. */
        data class Range(val min: Double, val max: Double) : Amount
    }

    private fun isUsableNumber(value: Double?): Boolean =
        value != null && value.isFinite() && value > 0.0

    /**
     * The amount a slot records, or null when its shape is not one of the two
     * the contract allows.
     *
     * A scalar carrying bounds is rejected rather than read as its scalar: the
     * row asserts both "exactly 620" and "anywhere in 560-700", and a reader
     * that picks one is guessing which half of a contradiction was meant.
     */
    fun amountOf(slot: StoredChemicalDefaultRate): Amount? {
        val hasScalar = slot.value != null
        val hasMin = slot.minValue != null
        val hasMax = slot.maxValue != null
        return when {
            hasScalar && !hasMin && !hasMax ->
                if (isUsableNumber(slot.value)) Amount.Scalar(slot.value!!) else null
            !hasScalar && hasMin && hasMax -> {
                val min = slot.minValue!!
                val max = slot.maxValue!!
                // An inverted band is not a narrow band, it is a corrupt one.
                if (isUsableNumber(min) && isUsableNumber(max) && min <= max) {
                    Amount.Range(min, max)
                } else {
                    null
                }
            }
            // Everything else: a lone bound, an empty row, or both shapes at once.
            else -> null
        }
    }

    /** A stored slot that passed every rule, with its amount already resolved. */
    data class ValidSlot(
        val basis: ChemicalDefaultRateBasis,
        /** Canonical spelling of the LABEL rate's unit — never the pack unit. */
        val unit: String,
        val amount: Amount,
        val slot: StoredChemicalDefaultRate,
    ) {
        /** The confirmed dose, or null when this slot records an unnarrowed band. */
        val scalar: Double? get() = (amount as? Amount.Scalar)?.value

        /** The confirmed band, or null when this slot records a single dose. */
        val range: Amount.Range? get() = amount as? Amount.Range

        /**
         * True when the operator typed this amount rather than reading it off a
         * registered direction. Drives wording — "User-confirmed" rather than
         * "Verified official rate" — and never weakens validation.
         */
        val isManualEntry: Boolean get() = slot.isManualEntry

        /** True when a human settled on this amount, whatever its origin. */
        val isConfirmedByOperator: Boolean get() = slot.isConfirmedByOperator
    }

    /**
     * The slot stored on [basis], or null when there is none the contract can
     * believe.
     *
     * Checked in full, because each rule catches a different way a row stops
     * meaning what it says:
     *
     * ```text
     * root version          a document written to a contract this build cannot read
     * containing basis      a per-100 L rate filed under per-hectare, applied per hectare
     * option key identity   a key no minter produced, so no client can match it
     * rate ids present      a default citing nothing can never be shown to be current
     * rate id identity      a UUID cites a row, not a printed direction
     * supported unit        an amount nothing can display or cost
     * source vocabulary     provenance outside the closed set is unattributable
     * amount shape (D3)     scalar XOR range, both finite and positive
     * ```
     */
    fun validSlot(
        defaults: StoredChemicalDefaultRates?,
        basis: ChemicalDefaultRateBasis,
    ): ValidSlot? {
        if (defaults == null) return null
        if (defaults.version != StoredChemicalDefaultRates.DEFAULT_RATES_VERSION) return null
        val slot = defaults.slot(basis) ?: return null

        val optionKey = slot.optionKey.trim()
        val rateIds = slot.rateIds.map { it.trim() }

        if (slot.isManualEntry) {
            // A manual rate has no official identity, and that is the point.
            // Demanding one here is exactly what made a user-confirmed
            // 2–3 L/100 L read as "Rate confirmation required" forever.
            //
            // The absence is CHECKED rather than merely tolerated: a row
            // claiming to be manual while carrying a citation is not a manual
            // rate expressed oddly, it is a row whose provenance contradicts
            // itself, and believing either half would lend label authority to
            // something a human typed.
            if (optionKey.isNotEmpty()) return null
            if (rateIds.any { it.isNotEmpty() }) return null
        } else {
            if (!optionKey.startsWith(OPTION_KEY_PREFIX)) return null
            if (optionKey.length <= OPTION_KEY_PREFIX.length) return null

            // Trimmed rather than filtered: a blank entry in the citation list
            // is a malformed row, not a row with one fewer citation.
            if (rateIds.isEmpty()) return null
            if (rateIds.any { it.length <= RATE_ID_PREFIX.length || !it.startsWith(RATE_ID_PREFIX) }) {
                return null
            }
        }

        if (slot.basis.trim() != basis.raw) return null

        val unit = canonicalUnit(slot.unit) ?: return null

        val source = slot.source.trim()
        val knownSource = source == StoredChemicalDefaultRate.SOURCE_OPERATOR ||
            source == StoredChemicalDefaultRate.SOURCE_RECOMMENDED
        if (!knownSource) return null

        val amount = amountOf(slot) ?: return null
        return ValidSlot(basis = basis, unit = unit, amount = amount, slot = slot)
    }

    /**
     * The CONFIRMED dose stored on [basis], or null when there is none.
     *
     * The scalar shape is the only confirmed amount: a slot still holding a
     * band is a decision that was never finished, and its bounds are what the
     * label permits rather than what this vineyard pours.
     */
    fun confirmedScalar(
        defaults: StoredChemicalDefaultRates?,
        basis: ChemicalDefaultRateBasis,
    ): ValidSlot? = validSlot(defaults, basis)?.takeIf { it.scalar != null }

    /** Every basis carrying a confirmed dose, per-hectare first. */
    fun confirmedScalars(defaults: StoredChemicalDefaultRates?): List<ValidSlot> =
        listOfNotNull(
            confirmedScalar(defaults, ChemicalDefaultRateBasis.PER_HECTARE),
            confirmedScalar(defaults, ChemicalDefaultRateBasis.PER_100_LITRES),
        )

    /** Every basis carrying a believable slot, per-hectare first. */
    fun validSlots(defaults: StoredChemicalDefaultRates?): List<ValidSlot> =
        listOfNotNull(
            validSlot(defaults, ChemicalDefaultRateBasis.PER_HECTARE),
            validSlot(defaults, ChemicalDefaultRateBasis.PER_100_LITRES),
        )

    /**
     * Every basis carrying an operator-confirmed amount — scalar OR range.
     *
     * A confirmed range IS a real decision: the operator has said "this
     * product, this band, on this basis". What it does not settle is the dose
     * for one particular tank, and that is asked for when the spray is built
     * rather than guessed here.
     */
    fun confirmedSlots(defaults: StoredChemicalDefaultRates?): List<ValidSlot> =
        validSlots(defaults).filter { it.isConfirmedByOperator }

    /**
     * Whether [value] lies inside a confirmed band, inclusive of both ends.
     *
     * Inclusive because a label authorises its own bounds: 2 and 3 are both
     * legal doses of a `2–3 L/100 L` direction.
     */
    fun isWithinRange(value: Double, min: Double, max: Double): Boolean =
        value.isFinite() && min.isFinite() && max.isFinite() && value >= min && value <= max
}
